src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -172,15 +172,31 @@ protected SourcePropertySymbolBase(
             {
                 Debug.Assert(!IsIndexer);
                 string fieldName = GeneratedNames.MakeBackingFieldName(_name);
-                BackingField = new SynthesizedBackingFieldSymbol(this,
-                                                                      fieldName,
-                                                                      // Synthesized backing field for 'field' should not be marked 'initonly'
-                                                                      // since the field might be modified in the get accessor.
-                                                                      // PROTOTYPE: Should the backing field be 'initonly' when the containing
-                                                                      // type, property, or accessor is declared 'readonly'?
-                                                                      isReadOnly: !usesFieldKeyword && ((hasGetAccessor && !hasSetAccessor) || isInitOnly),
-                                                                      this.IsStatic,
-                                                                      hasInitializer);
+
+                // The backing field is readonly if any of the following holds:
+                // - The containing type is declared readonly and the property is an instance property.
+                bool isReadOnly;
+                if (!IsStatic && containingType.IsReadOnly)
+                {
+                    isReadOnly = true;
+                }
+                // - The property is declared readonly.
+                else if (HasReadOnlyModifier)
+                {
+                    isReadOnly = true;
+                }
+                // - The property has no set accessor (but may have an init accessor) and
+                // the get accessor, if any, is automatically implemented.
+                else if ((!hasSetAccessor || isInitOnly) && (!hasGetAccessor || hasAutoPropertyGet))
```

Comment:
Should the field also be readonly if all manually implemented `get` and `set` accessors are declared `readonly`?

Reply:
It's not valid to declare both accessors as readonly, you'd have to declare the containing property as readonly instead, which would make the field readonly per these rules.

Reply:
Declaring both as `readonly` is an error, but we may want to treat the backing field as `readonly` in the following case:
```csharp
struct S
{
    object P { get; readonly set { } }
}
```

Reply:
I think a more compelling example is:
```cs
struct S
{
    object P { readonly get => field; init; }
}
```

Reply:
FWIW, I can't think of a case where things are made worse by the backing field not being readonly, in cases involving the accessors being marked readonly but not the containing type or property.

Reply:
I've added a PROTOTYPE comment and a test for now.